tracer/src/Datadog.Trace/Agent/DiscoveryService/DiscoveryService.cs

```diff
@@ -15,58 +15,89 @@ namespace Datadog.Trace.Agent.DiscoveryService
 
-        public string ProbeConfigurationEndpoint { get; private set; }
+        public string ConfigurationEndpoint { get; private set; }
 
         public string DebuggerEndpoint { get; private set; }
 
         public string AgentVersion { get; private set; }
 
-        public static DiscoveryService Create(IConfigurationSource configurationSource, IApiRequestFactory apiRequestFactory)
+        public static DiscoveryService Create(IApiRequestFactory apiRequestFactory)
         {
-            var exporterSettings = new ExporterSettings(configurationSource);
-            return new DiscoveryService(exporterSettings.AgentUri, apiRequestFactory);
+            lock (GlobalLock)
+            {
+                return Instance ??= new DiscoveryService(apiRequestFactory);
+            }
         }
 
         public async Task<bool> DiscoverAsync()
         {
-            try
+            // retry up to 5 times with exponential back-off
+            var retryLimit = 3;
+            var retryCount = 1;
+            var sleepDuration = 500; // in milliseconds
+
+            while (!_cancellationSource.IsCancellationRequested)
             {
-                var api = _apiRequestFactory.Create(new Uri($"{_agentUri}/info"));
-                using var response = await api.GetAsync().ConfigureAwait(false);
-                if (response.StatusCode != 200)
+                try
                 {
-                    Log.Error("Failed to discover services");
-                    return false;
+                    var uri = _apiRequestFactory.GetEndpoint("info");
+                    var api = _apiRequestFactory.Create(uri);
+
+                    using var response = await api.GetAsync().ConfigureAwait(false);
+                    if (response.StatusCode == 200)
+                    {
+                        var content = await response.ReadAsStringAsync().ConfigureAwait(false);
+                        ProcessDiscoveryResponse(content);
+
+                        _cancellationSource = null;
+                        return true;
+                    }
+
+                    Log.Warning("Failed to discover services");
+                }
+                catch (Exception exception)
+                {
+                    Log.Warning(exception, "Failed to discover services");
                 }
 
-                var content = await response.ReadAsStringAsync().ConfigureAwait(false);
-                ProcessDiscoveryResponse(content);
+                try
+                {
+                    await Task.Delay(sleepDuration, _cancellationSource.Token).ConfigureAwait(false);
+                }
+                catch (TaskCanceledException)
+                {
+                    return false;
+                }
 
-                return true;
-            }
-            catch (Exception exception)
-            {
-                Log.Error(exception, "Failed to discover services");
-                return false;
+                if (retryCount <= retryLimit)
```

Comment:
Maybe the name is unfortunate, but retryLimit just limiting how many times the exponential time grows. Attempt to connect to remove-config should continue until it manages. Otherwise the tracer might get configurations only at the next start.

Reply:
I've made it more clear at 908e120400d8db1dd36cfed0e4d55c2886516c8e